Fixes numeric overflow for the AVG function

tkroman opened the pull request:
Uses algorithm from https://en.wikipedia.org/wiki/Moving_average#Cumulative_moving_average.

I noticed this when I tried to run some statistics on my working dataset to find an appropriate block size (BTW, is it sensible to search for optimization possibilities in this direction?). 

Nodes have `attributes` property, which is basically a serialized map. It's length varies, but on average it's something like 175-200 characters, and the count of nodes is `14983680`, so the overflow happens:

```
scala> res5 * 175
res6: Int = -1672823296

scala> BigInt(res5) * 175
res7: scala.math.BigInt = 2622144000
```

Here's the screenshot that demonstrates the issue:

![Issue](https://i.imgur.com/ie8t8Lk.png)

As it's shown, `BigInt` could be used in order to not complicate the algorithm, but since CMA algorithm works, I decided to go with it.

tkroman committed:
Using Kahan's algorithm to compute AvgFunctions. Still losing precision.

thobe commented:
I really like this change. @tkroman if we could get you to agree to our contributor agreement we could merge this change.

See details here: http://neo4j.com/docs/stable/cla.html

tkroman commented:
@thobe, done. Glad I could help. It's weird that something fails though, since I ran all the Cypher tests and they are green.

thobe commented:
The CI failure is the verification of you being on the contributor list. Since you sent in the signed CLA I have added you and re-triggered the build.

tkroman commented:
@thobe, this may still fail then since I might have given you a different email from the one that is used here on Github. If that happens, here's my github email: rmn.tk.ml@gmail.com

tkroman commented:
@thobe, I decided to add a randomized test and found out this:

``` scala
test("onRandomList") {
  val N: Int = 10 * 1000
  val xs: Array[Long] = Array.fill(N)(Random.nextLong())
  val trueSum: BigInt = xs.sum
  val trueAvg: BigDecimal = BigDecimal(trueSum) / N

  val result = aggregateOn(xs: _*)

  result should equal(trueAvg.doubleValue())
}
```

It fails with `7.377319733210304E16 did not equal 7.3773197332103296E16`. This is because we lose precision with different speed in these two cases.

So, after all, maybe it's better to make use of `BigDecimal` which will be definitely slower, but guarantee some precision?

thobe commented:
@tkroman If the end result is going to be a double anyhow, I'm not sure going to BigDecimal is worth the overhead. Quite possibly we could use the [Kahan summation algorithm](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) to compensate for the error.

It would of course be very interesting to measure the performance difference between the different summation algorithms (straight up double Cumulative Moving Average, BigDecimal sum divided by count, and Kahan Summated Cumulative Moving Average).

thobe commented:
@tkroman the most recent build failure seems to stem from a fault in the revision you based your branch on. If you could rebase the pull request that would be great.

tkroman commented:
@thobe, the problem is (I believe) due to the precision loss during the division part (`(x(n + 1) - cma) / (n + 1)`), so the Kahan's algorithm doesn't help much (actually, I just tried it and the precision wen down another order). 
I'll poke around this a little more, let's put the merge on gold for now.

Upd. Kahan's method works just fine, but is still doesn't give the perfect result. Now 3-5 times out of 10 expected and actual values are equals, but I added another naively calculated average based on doubles and here's what I get:

``` scala
test("onRandomList") {
  val N: Int = 10 * 1000
  val xs: Array[Long] = Array.fill(N)(Random.nextLong())
  val trueSum: BigInt = xs.map(x => BigInt(x)).sum
  val trueAvg: BigDecimal = BigDecimal(trueSum, MathContext.DECIMAL64) / N
  val trueSumDouble = xs.map(x => x.toDouble).sum
  val trueAvgDouble = trueSumDouble / N

  val result = aggregateOn(xs: _*)

  println(s"BigDecimal: $trueAvg")
  println(s"Double: $trueAvgDouble")
  println(s"CMA+Kahan: $result")

  result should equal(trueAvg.doubleValue())
}
```

```
// Not equal, fails
BigDecimal: -7.13087205201199E+15
Double:     -7.130872052011962E15
CMA+Kahan:  -7.130872052011983E15
```

```
// Another run, equal, green tests
BigDecimal: -2.136840130944712E+16
Double:     -2.1368401309447156E16
CMA+Kahan:  -2.136840130944712E16
```

What do you think?

thobe commented:
Sorry, I missed that you had updated your comment. Github doesn't send out notifications when existing comments are changed. I hope you haven't waited for a reply for too long.

Great that you got Kahan's algorithm working. I was not expecting it to yield perfect results, only better results than using straight up CMA or computing the average as a double.

From your results it looks like the CMA+Kahan version you have yields a result that seems to be close enough to the BigDecimal version.
The remaining question is how they perform compared to one another. If the BigDecimal version has a similar performance profile to the CMA+Kahan version it might be preferable to just use that version, given that it is simpler.

I'll run some benchmarks based on your PR and get back to you.

tkroman commented:
@thobe, oops, didn't know that. Yes, Kahan yields much better results that the first version. I asked around at the university and talked to some of my colleagues - nobody seems to know something better, so I believe it's just a matter of how much of a precision loss this function can afford and how much is it worth.
I'll keep an eye on this thread, thanks for getting back at me.

cleishm commented:
@thobe make sure your benchmarks also evaluate space overheads (memory allocations), as I'd imagine BigDecimal would have a much larger impact in that area.

thobe commented:
@cleishm yes. although it's one of those cases where I'd expect hotspot to be able to do a decent job on eliminating the allocations, and whenever you start measuring allocations, all such optimisations go out the window. But I'll instead do long runs and measure the effects of allocation, i.e. garbage collection.

tkroman commented:
@thobe, by the way, the same (non-zero overflow probability) applies to the `SUM` function, so if your benchmarks will demonstrate that the overhead of using arbitraty-precision arithmetic is more or less insignificant, `SUM`'s implementation should be fixed as well.

systay commented:
bump

thobe commented:
I've run some benchmarks: https://gist.github.com/thobe/004fae93b0e2d519afc3

Kahan + CMA comes out over 50 times faster than using BigDecimal,
but almost 6,5 times slower than the simple algorithm of using single sum and then divide at the end.

It is also evident that the BigDecimal version has an unacceptably high object churn rate.

These are micro benchmarks, measuring only the essence of the algorithm. In reality there is a bit of overhead that gets in between. I believe the overhead that gets added in reality will:
- Make the difference between Kahan + CMA and our current algorithm _less evident_.
- Make it harder for the JVM to optimise away allocations made by the BigDecimal version, making that perform even worse in comparison.

tkroman commented:
Wow! The difference between CMA and simple division is unexpectedly impressive.

What are your thoughts on SUM? While using double as a default accumulation type will only increase the range by some amount, using BigDecimal/BigInt seems like a huuge slow-down. Maybe docs should just mention the possibility of overflow more explicitly :)

thobe commented:
Ok, so let's look at some benchmarks for just doing sums, using the corresponding algorithms:

```
Benchmark                                                    Mode  Cnt         Score        Error   Units
SumBenchmark.BigDecimalSum                                  thrpt   15        26.038 ±      0.426   ops/s
SumBenchmark.BigDecimalSum:gc.alloc.rate                    thrpt   15      2065.118 ±     33.794  MB/sec
SumBenchmark.BigDecimalSum:gc.alloc.rate.norm               thrpt   15  83166779.853 ±    904.443    B/op
SumBenchmark.BigDecimalSum:gc.count                         thrpt   15      4024.000               counts
SumBenchmark.BigDecimalSum:gc.time                          thrpt   15      2101.000                   ms

SumBenchmark.doubleSum                                      thrpt   15      8579.405 ±     30.066   ops/s
SumBenchmark.doubleSum:gc.alloc.rate                        thrpt   15        ≈ 10⁻⁵               MB/sec
SumBenchmark.doubleSum:gc.alloc.rate.norm                   thrpt   15         0.004 ±      0.001    B/op
SumBenchmark.doubleSum:gc.count                             thrpt   15           ≈ 0               counts

SumBenchmark.kahanSum                                       thrpt   15      2147.236 ±     12.482   ops/s
SumBenchmark.kahanSum:gc.alloc.rate                         thrpt   15        ≈ 10⁻⁵               MB/sec
SumBenchmark.kahanSum:gc.alloc.rate.norm                    thrpt   15         0.014 ±      0.001    B/op
SumBenchmark.kahanSum:gc.count                              thrpt   15           ≈ 0               counts
```

Here we see that the simple summation implementation is almost 4 times faster than the Kahan Summation algorithm. Comparing this to the 6.5 times difference we saw in the `AvgBenchmark`, it seems to make sense, since we were performing an additional subtraction and an additional division per iteration in that benchmark compared to what we do here.
The simple algorithm in AvgBenchmark was almost as fast as the one in SumBechmark, which also makes sense, since the integer addition that differentiates them is extremely cheap.

What is of course most interesting is how the BigDecimal version stands out, both by being substantially slower than the other two versions, but also by how much memory it needs to allocate. It has an 80x allocation overhead. The input data I'm summarising is 1MiB and for every time it summarises that it allocates approximately 80MiB.

My take on this is that as long as the answer is to be presented as a `double`, using Kahan's algorithm to minimise the summation error seems to be enough, the added benefit of a slight improvement in accuracy (note that since we convert to double it is still not exact, just slightly more accurate) we would get from BigDecimal does not justify the high cost it comes with.
If, however, we were to present the result in the form of a BigDecimal, where we could keep the value completely exact, the cost is probably worth it, since exactness would be the priority.

This argument would apply equally to both `sum` and `avg`.

tkroman committed:
Refactored Kahan's summation method into a separate class ; using it for SUM as well as AVG

tkroman committed:
Upodated docs for AVG/SUM/KahanSum

tkroman commented:
So, I updated both SUM and AVG.
I also changed `SumFunctionTest` so it will always expect `Double`, but perhaps there should be a check in the `result` call that will coerce `sum` to `Integer` in case it is safe to.

thobe commented:
I think `sum` should accumulate as an (exact) integer as long as all the values being summated over are integers.

thobe commented:
I spotted an error in my benchmarked code. After updating, the `kahanCMA` comes out a bit slower, 10.4 times slower than the version that just uses `double`.

I also added a version doing straight up CMA for comparison. This comes out only 21% faster than the `kahanCMA`, which illustrates that most of the cost comes from CMA, not the Kahan part. Which leads me to conclude that if we are to do CMA (which I think we should) we should definitely use it in conjunction with Kahan's summation algorithm.

tkroman committed:
Made SUM and AVG use gradually type-widening algorithms

tkroman commented:
So, I tinkered around this place again. I started with a simple procedural approach with a couple of cases, but code for handling type widening corner-cases became bloated, so I made a couple of classes, but if that's unnecessary and hits the performance, I'll drop this OOP stuff.
I also added a couple of tests to check for correctness. They demonstrate the basic idea: as long as it can, `Int` sum stays an `Int`, then widens to `Long`, then - to `Double`.

Mats-SX commented:
Hey @tkroman, could you please rebase this PR towards the 3.0 branch? We're willing to merge it in, but not in the 2.3 branch.

(PS: Since the path has changed name for many of the files you've edited, you might want to set the merge.renamelimit config on git to a high value.)

tkroman commented:
Hi @Mats-SX, sorry for the delay, been busy with all the work stuff. I made my branch fit to the 3.0 one, so I guess I'm closing this one.
